Port Singleblock Recipemap Machines + Generators to MUI2

ghzdude opened the pull request:
## What
Ports every singleblock Recipemap machine and generator, along with steam machines/boilers, to MUI2
Built off of the `sb-recipe-map-mui` branch

## Implementation Details
adds `usesMui2()` to SimpleGeneratorMTE, SimpleMachineMTE, and SteamMTE to defer to recipe map ui's `usesMui2()`
steam boilers ui is manually constructed, and therefore ported directly to Mui2
slot overlays for mui2 are stored in an Int2Obj map, with the key being the slot index
adds `RecipeProgressWidget` to allow for JEI searching
adds `RecipeMapUIBuilder` for building the mui2 slot overlays
deprecate old mui code and methods

## Outcome
the last set of uis should be ported to mui2

serenibyss committed:
stuff

serenibyss committed:
group slots, fix shift clicking

serenibyss committed:
clean up toggle buttons

serenibyss committed:
cleanups

serenibyss committed:
2.4 fix

serenibyss committed:
Logo as part of GTGuiTheme

serenibyss committed:
clickable progress bar (partial)

serenibyss committed:
finish previous commit

serenibyss committed:
fix progress bar sync

serenibyss committed:
fix button/toggle button theme builder calls

ghzdude committed:
fix rebase

ghzdude committed:
move mui2 related methods into builder
use builder consumer

ghzdude committed:
use unary operator instead

ghzdude committed:
port arc, assembler, autoclave

ghzdude committed:
spotless

ghzdude committed:
simplify builder to modify map ui directly
add a few more overloads

ghzdude committed:
move back to consumer

ghzdude committed:
add comment clarifying null supression
update as many singleblock machines to mui2

ghzdude committed:
set mui2 for map ui
spotless

ghzdude committed:
use progress value directly in widget
fix special texture
use GTFluidSlot instead of FluidSlot
use IDrawable.of() for slot overlays

ghzdude committed:
rework slot overlay methods
set max inputs before ui build
split overlay maps into Int2Obj maps
pull pos out of make slot methods

ghzdude committed:
port steam machines to mui2
simplify usesMui2 checks

ghzdude committed:
actually port steam machines to mui2
add todos for fluid slot

ghzdude committed:
port singleblock generators to mui2

ghzdude committed:
spotless

ghzdude committed:
improve/cleanup existing mui2 code
add todo

M-W-K commented on src/main/java/gregtech/api/recipes/RecipeMapBuilder.java:
```diff
@@ -223,24 +272,36 @@ public RecipeMapBuilder(@NotNull String unlocalizedName, @NotNull B defaultRecip
-        for (var entry : slotOverlays.byte2ObjectEntrySet()) {
-            ui.setSlotOverlay(entry.getByteKey(), entry.getValue());
+        if (usesMui2) {
+            // if "usesMui2" is true, then "mapUIBuilder" is not null
```
Why not just nullcheck mapUIBuilder?

M-W-K commented on src/main/java/gregtech/api/recipes/RecipeMapBuilder.java:
```diff
@@ -141,79 +156,113 @@ public RecipeMapBuilder(@NotNull String unlocalizedName, @NotNull B defaultRecip
      * @return this
      */
+    @Deprecated
+    @ApiStatus.ScheduledForRemoval(inVersion = "2.9")
```
Isn't our next release going to be 2.9? Probably ought to schedule removal for 2.10 or so.

M-W-K reviewed (commented):
Did a first look through the code, did not test ingame.

ghzdude committed:
spotless

ghzdude committed:
null check ui builder and simplify

Zorbatron committed:
Actually merge master in properly :facepalm:

ALongStringOfNumbers reviewed (commented):
The fluid amount overlay on slots seems to be slightly adjusted from what it was before. Was that something touched in this PR, or am I just crazy?

ghzdude committed:
remove old ui code from SteamBoiler and extending classes

ghzdude committed:
fix SteamLavaBoiler lava fluid slot

ghzdude committed:
use `slotOverlay()` directly

ghzdude committed:
rename mui2 special drawable

ghzdude committed:
rename mui2 special drawable and method
address todo

ghzdude commented on src/main/java/gregtech/api/recipes/ui/RecipeMapUI.java:
```diff
@@ -223,15 +267,15 @@ protected void addInventorySlotGroup(@NotNull ModularUI.Builder builder,
                 for (int i = 0; i < fluidInputsCount; i++) {
                     int y = startInputsY + 18 * i;
-                    addSlot(builder, startSpecX, y, i, itemHandler, fluidHandler, !invertFluids, isOutputs);
+                    addSlot(builder, startSpecX, y, i, itemHandler, fluidHandler, true, isOutputs);
```
intellij tells me that it's always true for some reason

ghzdude committed:
remove irrelevant todo

ghzdude committed:
remove commented code

ghzdude committed:
actually sync the sync value and remove todo

ghzdude committed:
spotless

ghzdude committed:
initial rework of `addInventorySlotGroup()` for mui2

ghzdude committed:
more work on the rewrite
add debug names
clarify some variable names
calculate the grid size for both items and fluids

ALongStringOfNumbers commented:
Also, we think that we should implement the side column, like in multiblocks and hatches, for clearing out some space in the GUI. See the discussion starting here: https://discord.com/channels/701354865217110096/904846945095385138/1408285204795359292

ghzdude committed:
use modular panel instead of redundant parent widget
use qualified references for old gui class references

ghzdude committed:
rework mostly done

ghzdude committed:
clean + spotless

ghzdude committed:
cleanup pt 2
improve overlay map

ghzdude committed:
fix no energy indicator
fix generator ui

ghzdude committed:
replace special texture with a widget consumer

ghzdude committed:
fix special textures
remove `+ 9`

ghzdude committed:
remove if condition for simple machines

ghzdude committed:
increase progress widget margin

ghzdude committed:
improve centering slots around progress widget

ghzdude committed:
improve item/fluid groups
move special machine widgets around

ghzdude committed:
make `setSize()` a builder method
construct panel in map ui
adjust height to remove empty space

ghzdude committed:
create CalculatedGrid to simplify and cleanup logic

ghzdude committed:
fix logo in SimpleGeneratorMetaTileEntity
add heat tooltip + lang in SteamBoiler
add todo for SteamMetaTileEntity

ghzdude committed:
remove unused field

ghzdude committed:
fix rebase

ghzdude committed:
add gt logo to steam machines
shrink backgrounds to 16x16
remove logo from backgrounds

ghzdude committed:
move steam/energy indicators to overlay progress widget
update indicator textures with black border
fix steam machines being too wide
add note

bruberu reviewed (approved):
The actual features work, and I don't see any massive code smells.

ghzdude committed:
fix indicator replacing forge hammer overlay
move indicator back down below progress widget

ghzdude committed:
Revert "update indicator textures with black border"

ghzdude committed:
rework panel constructing to use a builder
move `setSize()` into builder
move yoffset into builder
add a method for widgets that need yoffset
use handler references in CalculatedGrid
simplify a few method signatures

ghzdude committed:
move several ui related methods into PanelBuilder
rename Builder -> PanelBuilder
create empty CalculatedGrid as placeholder
make methods protected for override (if desired)

ghzdude committed:
increase height to give energy/steam indicator more room
